File: vonage_video_bridge/android/src/main/kotlin/com/example/vonage_video_bridge/VonageVideoBridgePlugin.kt
package com.example.vonage_video_bridge

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.opentok.android.BaseVideoRenderer
import com.opentok.android.OpentokError
import com.opentok.android.Publisher
import com.opentok.android.PublisherKit
import com.opentok.android.Session
import com.opentok.android.Stream
import com.opentok.android.Subscriber
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class VonageVideoBridgePlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler, ActivityAware {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var controller: VonageVideoController

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        controller = VonageVideoController(binding.applicationContext)
        methodChannel = MethodChannel(binding.binaryMessenger, "vonage_video_bridge/methods")
        eventChannel = EventChannel(binding.binaryMessenger, "vonage_video_bridge/events")
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
        binding.platformViewRegistry.registerViewFactory(
            "vonage_video_bridge/view",
            VonageVideoPlatformViewFactory(controller)
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        controller.disconnect()
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        controller.activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        controller.activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        controller.activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        controller.activity = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        controller.eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        controller.eventSink = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "connect" -> {
                    val applicationId = call.argument<String>("applicationId") ?: error("applicationId is required")
                    val sessionId = call.argument<String>("sessionId") ?: error("sessionId is required")
                    val token = call.argument<String>("token") ?: error("token is required")
                    val publisherName = call.argument<String>("publisherName") ?: "Flutter"
                    controller.connect(applicationId, sessionId, token, publisherName)
                    result.success(null)
                }
                "publish" -> {
                    controller.publish()
                    result.success(null)
                }
                "setAudioEnabled" -> {
                    controller.setAudioEnabled(call.arguments as Boolean)
                    result.success(null)
                }
                "setVideoEnabled" -> {
                    controller.setVideoEnabled(call.arguments as Boolean)
                    result.success(null)
                }
                "switchCamera" -> {
                    controller.switchCamera()
                    result.success(null)
                }
                "disconnect" -> {
                    controller.disconnect()
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        } catch (e: Throwable) {
            result.error("VONAGE_VIDEO_ERROR", e.message, null)
        }
    }
}

class VonageVideoController(
    private val context: Context
) : Session.SessionListener, PublisherKit.PublisherListener {
    var activity: Activity? = null
    var eventSink: EventChannel.EventSink? = null

    private var session: Session? = null
    private var publisher: Publisher? = null
    private var subscriber: Subscriber? = null
    private var container: FrameLayout? = null

    fun attach(root: FrameLayout) {
        container = root
        renderViews()
    }

    fun detach(root: FrameLayout) {
        if (container === root) {
            container = null
        }
    }

    fun connect(applicationId: String, sessionId: String, token: String, publisherName: String) {
        ensurePermissions()
        disconnect()
        publisher = Publisher.Builder(context).name(publisherName).build().also {
            it.setPublisherListener(this)
            it.renderer?.setStyle(BaseVideoRenderer.STYLE_VIDEO_SCALE, BaseVideoRenderer.STYLE_VIDEO_FILL)
        }
        session = Session.Builder(context, applicationId, sessionId).build().also {
            it.setSessionListener(this)
            it.connect(token)
        }
        event("connecting")
        renderViews()
    }

    fun publish() {
        val activeSession = session ?: error("Session is not connected")
        val activePublisher = publisher ?: error("Publisher is not ready")
        activeSession.publish(activePublisher)
        renderViews()
    }

    fun setAudioEnabled(enabled: Boolean) {
        publisher?.publishAudio = enabled
    }

    fun setVideoEnabled(enabled: Boolean) {
        publisher?.publishVideo = enabled
    }

    fun switchCamera() {
        publisher?.cycleCamera()
    }

    fun disconnect() {
        subscriber?.let { session?.unsubscribe(it) }
        publisher?.let { session?.unpublish(it) }
        session?.disconnect()
        subscriber = null
        publisher = null
        session = null
        container?.removeAllViews()
        event("disconnected")
    }

    override fun onConnected(session: Session) {
        event("connected")
        publish()
    }

    override fun onDisconnected(session: Session) {
        event("disconnected")
    }

    override fun onStreamReceived(session: Session, stream: Stream) {
        event("streamReceived", streamId = stream.streamId)
        subscriber = Subscriber.Builder(context, stream).build()
        session.subscribe(subscriber)
        renderViews()
    }

    override fun onStreamDropped(session: Session, stream: Stream) {
        event("streamDropped", streamId = stream.streamId)
        subscriber = null
        renderViews()
    }

    override fun onError(session: Session, opentokError: OpentokError) {
        event("error", opentokError.message)
    }

    override fun onStreamCreated(publisherKit: PublisherKit, stream: Stream) {
        event("publisherStreamCreated", streamId = stream.streamId)
    }

    override fun onStreamDestroyed(publisherKit: PublisherKit, stream: Stream) {
        event("publisherStreamDestroyed", streamId = stream.streamId)
    }

    override fun onError(publisherKit: PublisherKit, opentokError: OpentokError) {
        event("error", opentokError.message)
    }

    private fun renderViews() {
        val root = container ?: return
        root.removeAllViews()

        subscriber?.view?.let { remote ->
            addFullSize(root, remote)
        }

        publisher?.view?.let { local ->
            val size = (120 * root.resources.displayMetrics.density).toInt()
            val params = FrameLayout.LayoutParams(size, size).apply {
                gravity = Gravity.TOP or Gravity.END
                topMargin = 24
                rightMargin = 24
            }
            if (local.parent != null) {
                (local.parent as? FrameLayout)?.removeView(local)
            }
            root.addView(local, params)
        }
    }

    private fun addFullSize(root: FrameLayout, view: View) {
        if (view.parent != null) {
            (view.parent as? FrameLayout)?.removeView(view)
        }
        root.addView(
            view,
            FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            )
        )
    }

    private fun ensurePermissions() {
        val currentActivity = activity ?: return
        val missing = arrayOf(Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO)
            .filter {
                ContextCompat.checkSelfPermission(currentActivity, it) != PackageManager.PERMISSION_GRANTED
            }
            .toTypedArray()
        if (missing.isNotEmpty()) {
            ActivityCompat.requestPermissions(currentActivity, missing, 7001)
        }
    }

    private fun event(type: String, message: String? = null, streamId: String? = null) {
        eventSink?.success(
            mapOf(
                "type" to type,
                "message" to message,
                "streamId" to streamId
            )
        )
    }
}

File: vonage_video_bridge/ios/Classes/VonageVideoBridgePlugin.swift
import Flutter
import OpenTok
import UIKit

public class VonageVideoBridgePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private let controller = VonageVideoController()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = VonageVideoBridgePlugin()
    let methodChannel = FlutterMethodChannel(
      name: "vonage_video_bridge/methods",
      binaryMessenger: registrar.messenger()
    )
    let eventChannel = FlutterEventChannel(
      name: "vonage_video_bridge/events",
      binaryMessenger: registrar.messenger()
    )

    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
    registrar.register(
      VonageVideoPlatformViewFactory(controller: instance.controller),
      withId: "vonage_video_bridge/view"
    )
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    do {
      switch call.method {
      case "connect":
        guard let args = call.arguments as? [String: Any],
              let applicationId = args["applicationId"] as? String,
              let sessionId = args["sessionId"] as? String,
              let token = args["token"] as? String else {
          result(FlutterError(code: "BAD_ARGS", message: "applicationId, sessionId, and token are required", details: nil))
          return
        }
        let publisherName = args["publisherName"] as? String ?? "Flutter"
        try controller.connect(
          applicationId: applicationId,
          sessionId: sessionId,
          token: token,
          publisherName: publisherName
        )
        result(nil)
      case "publish":
        try controller.publish()
        result(nil)
      case "setAudioEnabled":
        controller.setAudioEnabled(call.arguments as? Bool ?? true)
        result(nil)
      case "setVideoEnabled":
        controller.setVideoEnabled(call.arguments as? Bool ?? true)
        result(nil)
      case "switchCamera":
        controller.switchCamera()
        result(nil)
      case "disconnect":
        controller.disconnect()
        result(nil)
      default:
        result(FlutterMethodNotImplemented)
      }
    } catch {
      result(FlutterError(code: "VONAGE_VIDEO_ERROR", message: error.localizedDescription, details: nil))
    }
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    controller.eventSink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    controller.eventSink = nil
    return nil
  }
}

final class VonageVideoController: NSObject, OTSessionDelegate, OTPublisherDelegate, OTSubscriberDelegate {
  var eventSink: FlutterEventSink?

  private var session: OTSession?
  private var publisher: OTPublisher?
  private var subscriber: OTSubscriber?
  private weak var container: UIView?

  func attach(_ view: UIView) {
    container = view
    renderViews()
  }

  func detach(_ view: UIView) {
    if container === view {
      container = nil
    }
  }

  func connect(applicationId: String, sessionId: String, token: String, publisherName: String) throws {
    disconnect()
    publisher = OTPublisher(delegate: self, name: publisherName)
    session = OTSession(apiKey: applicationId, sessionId: sessionId, delegate: self)

    var error: OTError?
    session?.connect(withToken: token, error: &error)
    if let error {
      throw error
    }
    sendEvent("connecting")
    renderViews()
  }

  func publish() throws {
    guard let session else {
      throw NSError(domain: "VonageVideoBridge", code: 1, userInfo: [NSLocalizedDescriptionKey: "Session is not connected"])
    }
    guard let publisher else {
      throw NSError(domain: "VonageVideoBridge", code: 2, userInfo: [NSLocalizedDescriptionKey: "Publisher is not ready"])
    }

    var error: OTError?
    session.publish(publisher, error: &error)
    if let error {
      throw error
    }
    renderViews()
  }

  func setAudioEnabled(_ enabled: Bool) {
    publisher?.publishAudio = enabled
  }

  func setVideoEnabled(_ enabled: Bool) {
    publisher?.publishVideo = enabled
  }

  func switchCamera() {
    guard let publisher else { return }
    publisher.cameraPosition = publisher.cameraPosition == .front ? .back : .front
  }

  func disconnect() {
    if let subscriber {
      var error: OTError?
      session?.unsubscribe(subscriber, error: &error)
    }
    if let publisher {
      var error: OTError?
      session?.unpublish(publisher, error: &error)
    }
    var error: OTError?
    session?.disconnect(&error)
    subscriber = nil
    publisher = nil
    session = nil
    container?.subviews.forEach { $0.removeFromSuperview() }
    sendEvent("disconnected")
  }

  func sessionDidConnect(_ session: OTSession) {
    sendEvent("connected")
    try? publish()
  }

  func sessionDidDisconnect(_ session: OTSession) {
    sendEvent("disconnected")
  }

  func session(_ session: OTSession, streamCreated stream: OTStream) {
    sendEvent("streamReceived", streamId: stream.streamId)
    subscriber = OTSubscriber(stream: stream, delegate: self)
    if let subscriber {
      var error: OTError?
      session.subscribe(subscriber, error: &error)
      if let error {
        sendEvent("error", message: error.localizedDescription)
      }
    }
    renderViews()
  }

  func session(_ session: OTSession, streamDestroyed stream: OTStream) {
    sendEvent("streamDropped", streamId: stream.streamId)
    subscriber = nil
    renderViews()
  }

  func session(_ session: OTSession, didFailWithError error: OTError) {
    sendEvent("error", message: error.localizedDescription)
  }

  func publisher(_ publisher: OTPublisherKit, streamCreated stream: OTStream) {
    sendEvent("publisherStreamCreated", streamId: stream.streamId)
  }

  func publisher(_ publisher: OTPublisherKit, streamDestroyed stream: OTStream) {
    sendEvent("publisherStreamDestroyed", streamId: stream.streamId)
  }

  func publisher(_ publisher: OTPublisherKit, didFailWithError error: OTError) {
    sendEvent("error", message: error.localizedDescription)
  }

  func subscriberDidConnect(toStream subscriberKit: OTSubscriberKit) {
    renderViews()
  }

  func subscriber(_ subscriber: OTSubscriberKit, didFailWithError error: OTError) {
    sendEvent("error", message: error.localizedDescription)
  }

  private func renderViews() {
    guard let container else { return }
    container.subviews.forEach { $0.removeFromSuperview() }

    if let remoteView = subscriber?.view {
      remoteView.frame = container.bounds
      remoteView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
      container.addSubview(remoteView)
    }

    if let localView = publisher?.view {
      let side: CGFloat = 120
      localView.frame = CGRect(
        x: container.bounds.width - side - 24,
        y: 24,
        width: side,
        height: side
      )
      localView.autoresizingMask = [.flexibleLeftMargin, .flexibleBottomMargin]
      container.addSubview(localView)
    }
  }

  private func sendEvent(_ type: String, message: String? = nil, streamId: String? = nil) {
    var payload: [String: Any] = ["type": type]
    if let message {
      payload["message"] = message
    }
    if let streamId {
      payload["streamId"] = streamId
    }
    eventSink?(payload)
  }
}
